src/libraries/System.Security.Cryptography/src/System/Security/Cryptography/Helpers.cs

```diff
@@ -468,5 +468,114 @@ internal static void ThrowIfPasswordContainsNullCharacter(string? password)
 
         internal static bool IsSlhDsaOid(string? oid) =>
             SlhDsaAlgorithm.GetAlgorithmFromOid(oid) is not null;
+
+        public delegate void SlhDsaPreHashActionCallback<T, TState>(Span<T> span, TState arg)
+            where TState : allows ref struct;
+
+        public delegate TResult SlhDsaPreHashFuncCallback<T, TState, TResult>(Span<T> span, TState arg)
+            where TState : allows ref struct;
+
+        internal static void SlhDsaPreHash<TState>(
+            ReadOnlySpan<byte> hash,
+            ReadOnlySpan<byte> context,
+            ReadOnlySpan<char> hashAlgorithmOid,
+            TState state,
+            SlhDsaPreHashActionCallback<byte, TState> callback)
+            where TState : allows ref struct
+        {
+            SlhDsaPreHash(
+                hash,
+                context,
+                hashAlgorithmOid,
+                CreateStackTuple(state, callback),
+                static (message, myState) =>
+                {
+                    myState.Item2(message, myState.Item1);
+                    return true;
+                });
+        }
+
+        internal static TResult SlhDsaPreHash<TState, TResult>(
+            ReadOnlySpan<byte> hash,
+            ReadOnlySpan<byte> context,
+            ReadOnlySpan<char> hashAlgorithmOid,
+            TState state,
+            SlhDsaPreHashFuncCallback<byte, TState, TResult> callback)
+            where TState : allows ref struct
+        {
+            // The OIDs for the algorithms above have max length 11. We'll just round up for a conservative initial estimate.
+            const int MaxEncodedOidLengthForCommonHashAlgorithms = 16;
+            AsnWriter writer = new AsnWriter(AsnEncodingRules.DER, MaxEncodedOidLengthForCommonHashAlgorithms);
+            writer.WriteObjectIdentifier(hashAlgorithmOid);
+
+            int encodedOidLength = writer.GetEncodedLength();
+            int messageLength =
+                1 +                 // Pre-hash encoding flag
+                1 +                 // Context length
+                context.Length +    // Context
+                encodedOidLength +  // OID
+                hash.Length;        // Hash
+
+            const int MessageLengthUpperBoundEstimate = 1 + 1 + 255 + 11 + (512 / 8);
```

Comment:
Once the hash is a XOF, there's no upper bound (so no 512/8), and there's no real limit to the number of arcs in an OID (and the number of bytes will be at minimum arcs-1), though AsnDecoder/AsnReader limit it to 64 for reasons.  So the 11 isn't sound, either.

If you just want a stackalloc heuristic, pick 256 or 512.